Fix - UPE Appearance on WC Blocks

achyuthajoy opened the pull request:
Fixes #4145 

#### Changes proposed in this Pull Request

Classic checkout automatically generates styles for the UPE payment form directly form the checkout page. However when using Blocks checkout this doesn't work because the appearance value was passed to the `PaymentElement` and appearance data was not generated since the selectors used on Blocks checkout are different.

This PR refactors the existing functions to generate UPE styles to allow style generation for Blocks checkout and adds the styles to the [Elements Appearance API](https://stripe.com/docs/stripe-js/appearance-api) instead of `PaymentElement`.

#### Testing instructions

* Enable UPE Payment methods & WC Blocks on the test site.
* On `develop` add a product to cart and go to classic checkout. Notice that the UPE Payment form styles match the input styles on the checkout page.
* Go to Blocks checkout and notice that the UPE payment form uses default styles and doesn't match the input styles on the page
* Checkout the branch `fix/upe-blocks-appearance` and rebuilt assets
* Go to classic checkout and confirm that the UPE payment form styles matches input elements on the page
* Go to Blocks checkout and confirm that the UPE payment form styles matches input elements on the page.

**Notes**
* The label transitions on Blocks checkout input fields are not passed to the payment form
* There is a workaround in place to rewrite `text-indent` used by WC Blocks to `padding`. WC Blocks introduced using `text-indent` on this [PR](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5947) and stripe doesn't support this property.

**Further Optimizations**

The styles for payment form on Blocks checkout like styles for invalid labels & validation messages can be optimized further on a different PR.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_

jessy-p commented:
Tested and working.
But the indent issue on the checkout blocks looks odd., should we also try to fix that before release?
<img width="506" alt="image" src="https://user-images.githubusercontent.com/32092402/165913099-b1b41b66-95ca-4344-aae6-3de937fc6d57.png">

jessy-p reviewed (commented):
Tested and works as described.
But the indent issue looks odd, should we try to fix that before releasing this?
<img width="506" alt="image" src="https://user-images.githubusercontent.com/32092402/165913485-ce038319-50bd-4c4c-8a02-b4ca3de401d5.png">

jbordonado commented:
I just did that and I have the same result.
I tested with WooCommerce Blocks 7.5.0, which I see was just announced as released so that's maybe fixed in this new version?

achyuthajoy commented:
@jbordonado I'm running the latest version of blocks and the padding issue still appears for me. I just pushed a workaround to fix the same. 

Also do you see same input styles as checkout blocks?

zmaglica commented:
I also have the same issue as @jessy-p . The code looks fine, and the bug is fixed. Yet, we will have to discuss how to fix this issue with paddings.

![image](https://user-images.githubusercontent.com/15913573/165919595-dade54d0-606c-4a20-afe4-aaf3b629950a.png)

achyuthajoy commented:
@zmaglica Can you pull the latest changes? I just pushed a workaround. 

You'll need to remove the existing transient from `wp_options` for the changes to take effect. - `_transient_wcpay_wc_blocks_upe_appearance`

zmaglica commented:
> _transient_wcpay_wc_blocks_upe_appearance

That did the trick. It is working fine on my end.

jbordonado commented:
> @jbordonado I'm running the latest version of blocks and the padding issue still appears for me. I just pushed a workaround to fix the same.
> 
> Also do you see same input styles as checkout blocks?

Not really, other inputs have border #50575e while payment fields have border #e6e6e6 for example.
Maybe something's wrong on my local setup, but I'm not doing anything in particular..

achyuthajoy commented:
@jbordonado Can you try building the plugin zip from this branch and trying it on a JN site?

FangedParakeet reviewed (commented):
I really like the approach here! The issue was to simply update the blocks element to use the `appearance` parameter correctly, but you've gone above and beyond by adding separate transients for blocks and classic checkouts, as well as computing styles from multiple different elements on the page (container, input, and invalid input). Nice one!

That being said, I feel we might be able to get even closer parity between the classic and blocks checkout. Here are a few screenshots that show the difference between the UPE appearance in classic and blocks.

By the way, the theme I like to test a lot with is Deli, because it's quite ugly, but has the most drastic changes to inputs and styles, so it's very good to test these settings!

![UPE classic](https://cdn-std.droplr.net/files/acc_1104206/0J16iL)
_UPE in classic checkout_

![UPE blocks](https://cdn-std.droplr.net/files/acc_1104206/VyghdA)
_UPE in blocks checkout_

You can see here that in the classic checkout, the UPE input picks up the correct border, background, and font styles; whereas the blocks UPE inputs do not pick up these styles. I wondered if that might be because you are pulling styles from the shipping fields in blocks and the billing fields in the classic checkout, but I haven't really tested this. 

Could you double-check the above and let me know if tweaking the inputs or where you place them might be able to get the styling a little closer to the classic checkout?

Also, for some reason, if I left the blocks checkout open for too long, I noticed this error in the console and my entire browser ground to a halt and eventually crashed. 😂 

> Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.

Can you double-check that as well? I'd presume that must be related to the `useEffect` code inside `upe-fields.js`--though, I was adding some code there at times to stop using cached styles, so perhaps I might have caused this to break myself.

I think this is a great approach; can you just have a little ponder on a few of the issues I've raised? I'll have another look once you think you've sorted everything out. Cheers!

achyuthajoy commented on client/checkout/upe-styles/index.js:
```diff
@@ -4,20 +4,204 @@
+		],
+	},
+	blocksCheckout: {
+		appendTarget: '#shipping.wc-block-components-address-form',
+		upeThemeInputSelector: '#shipping-first_name',
```
@FangedParakeet I believe you don't have shipping methods configured on WooCommerce. It looks like if shipping methods are present, WC Blocks displays shipping fields and if not, billing fields are displayed.

I haven't noticed this until now. Since blocks use react, the fields are added to the DOM on request, so we can't stick to a single selector. I'll add a check to find which fields are active on the page before generating the styles.

achyuthajoy commented on client/checkout/blocks/upe-fields.js:
```diff
@@ -263,6 +275,8 @@ const ConsumableWCPayFields = ( { api, ...props } ) => {
 	const options = {
 		clientSecret,
+		appearance,
+		fonts: fontRules,
```
I followed the same implementation style as UPE Classic Checkout

achyuthajoy commented:
@FangedParakeet Thanks for testing the changes thoroughly. 👍 

>You can see here that in the classic checkout, the UPE input picks up the correct border, background, and font styles; whereas the blocks UPE inputs do not pick up these styles. I wondered if that might be because you are pulling styles from the shipping fields in blocks and the billing fields in the classic checkout, but I haven't really tested this.
>
> Could you double-check the above and let me know if tweaking the inputs or where you place them might be able to get the styling a little closer to the classic checkout?

I've found out why this happened and mentioned it here - https://github.com/Automattic/woocommerce-payments/pull/4211#discussion_r862307929

> Also, for some reason, if I left the blocks checkout open for too long, I noticed this error in the console and my entire browser ground to a halt and eventually crashed. 😂
>
> ```Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.```
>
> Can you double-check that as well? I'd presume that must be related to the useEffect code inside upe-fields.js--though, I was adding some code there at times to stop using cached styles, so perhaps I might have caused this to break myself.

I've kept the page open for some time & I haven't noticed this behavior on my end. Since the style generation happens inside `useEffect` hook, it doesn't run with every render unless there is change in state. Could be related to manual changes while you were testing.

jbordonado commented:
> @jbordonado Can you try building the plugin zip from this branch and trying it on a JN site?

I was eventually able to reproduce the padding issue on a JN site and also on my local. But not at first. I started with the Storefront theme and the display was correct. I switched to Dali (as per the above comment) to have a look at another theme and the padding issue appeared. Weird.

After I pulled your latest changes and removed the transient, the padding issue was gone and the UPE payment form styles match input elements on the page, so I guess it's OK now!

FangedParakeet commented:
Thanks for these changes, @achyuthajoy! The code looks fine to me and the styles are looking great on both classic and blocks checkouts. 

However, when I first pulled these changes, the payment element wouldn't load and I started discovering JS errors in my console.

![Javascript booboo](https://cdn-std.droplr.net/files/acc_1104206/ALAntV)
_Payment element rules incorrectly formatted_

For some reason, the values for many of the rules properties were returning empty arrays instead of empty objects (see below). 

```js
// BAD
const appearance = {
   rules: {
      '.Input': [],
      '.Input:focus': [],
      ...
   }
}

// GOOD
const appearance = {
   rules: {
      '.Input': { },
      '.Input:focus': { },
      ...
   }
}

``` 

Switching the theme, which clears the appearance transient resolved this issue for me. However, this made me concerned that this PR might cause some kind of compatibility issues when we merge and roll it out to end-users. If users don't manually refresh their appearance transients, might this new code cause an error that would break their checkout entirely in blocks with the UPE enabled?

Do you think it would be possible to check for a malformed `appearance` object and, if discovered, to cause the plugin to refresh the transient? I think this would make things smoother, ensure that users experience your changes here, and would also make this PR easier for others to test. Let me know if you have any thoughts about this suggestion. Cheers!

achyuthajoy commented:
>For some reason, the values for many of the rules properties were returning empty arrays instead of empty objects

Hey @FangedParakeet! Nice find. 

I was able to figure out the problem(s).

1. The classic checkout UPE module was using jQuery to send the UPE appearance data which was sent as an array and then added to the transient. Here is the code that handles the request for Classic UPE - 
2. The blocks checkout UPE module uses a different JS function to send the UPE appearance data, which I've modified to send the data as JSON encoded.
3. Also, while saving the data to transient, I was decoding the appearance data as an array instead of object. This caused the mismatch in data fetched from the transient, if any of the rules are empty. If empty the rule returns an array, else it returns an object.

I've fixed the problem by:
1. Moving the request module from blocks checkout to utils and refactors classic upe to use the same request module. I've also updated the related tests & files.
2. The data is now sent to backend as stringified JSON and decoded as object before saving to the transient making the format same for classic & upe appearance.
3. I've also forced to delete the existing UPE appearance transients when WCPay version is updated.

It's good for another review :)

achyuthajoy commented:
Triggered an E2E test against this branch here - https://github.com/Automattic/woocommerce-payments/actions/runs/2265293107 to make sure payment processing works as expected.

FangedParakeet reviewed (approved):
Thanks for all the changes here! This is working splendidly for me now. Here's some proof with a few varying themes to display how in both the blocks and classic checkouts, the payment element now matches its neighbours perfectly and blends in seamlessly.

**Storefront**

![Classic storefront](https://cdn-std.droplr.net/files/acc_1104206/wz15UG)
_Classic checkout_

![Blocks storefront](https://cdn-std.droplr.net/files/acc_1104206/MU3ktD)
_Blocks checkout_

**Twenty Twenty-One**

![Classic 2021](https://cdn-std.droplr.net/files/acc_1104206/j0eteP)
_Classic checkout_

![Blocks 2021](https://cdn-std.droplr.net/files/acc_1104206/mr5fcw)
_Blocks checkout_

**Deli**

![Classic deli](https://cdn-std.droplr.net/files/acc_1104206/ONoORn)
_Classic checkout_

![Blocks deli](https://cdn-std.droplr.net/files/acc_1104206/T6IUg2)
_Blocks checkout_

I think this PR not only fixes the incorrect implementation of the appearance parameter for the Blocks checkout, but also massively improves the end-user experience for both checkouts, so great job! Everything here LGTM, so let's go. 🚢
